core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -0,0 +1,1870 @@
+      if (redeemScriptVec.size == 1) Some(redeemScriptVec.head.redeemScript)
+      else None
+
+    val scriptWitnessVec = getRecords(WitnessScriptKeyId)
+    val scriptWitnessOpt =
+      if (scriptWitnessVec.size == 1) {
+        Some(P2WSHWitnessV0(scriptWitnessVec.head.witnessScript))
+      } else if (output.scriptPubKey
+                   .isInstanceOf[P2WPKHWitnessSPKV0] || redeemScriptOpt.exists(
+                   _.isInstanceOf[P2WPKHWitnessSPKV0])) {
+        Some(P2WPKHWitnessV0(signer.publicKey))
+      } else {
+        None
+      }
+
+    val hashTypeVec = getRecords(SigHashTypeKeyId)
+    val hashType =
+      if (hashTypeVec.size == 1) hashTypeVec.head.hashType
+      else HashType.sigHashAll
+
+    BitcoinUTXOSpendingInfoSingle(outPoint,
+                                  output,
+                                  signer,
+                                  redeemScriptOpt,
+                                  scriptWitnessOpt,
+                                  hashType,
+                                  conditionalPath)
+  }
+
+  /**
+    * Check if this satisfies criteria for witness. If it does, delete the NonWitnessOrUnknownUTXO field
+    * This is useful for following reasons.
+    * 1. Compresses the size of the data
+    * 2. Allows for segwit only compatibility
+    * @param txIn The TransactionInput that this InputPSBTMap represents
+    * @return The compressed InputPSBTMap
+    */
+  def compressMap(txIn: TransactionInput): InputPSBTMap = {
+    if (isFinalized) {
+      this
+    } else {
+      val newElements = {
+        val nonWitUtxoVec =
+          this.getRecords(NonWitnessUTXOKeyId)
+        val witScriptVec =
+          this.getRecords(WitnessScriptKeyId)
+        if (nonWitUtxoVec.size == 1 && witScriptVec.size == 1) {
+          val nonWitUtxo = nonWitUtxoVec.head.transactionSpent
+          if (txIn.previousOutput.vout.toInt < nonWitUtxo.outputs.size) {
+            val out = nonWitUtxo.outputs(txIn.previousOutput.vout.toInt)
+            elements
+              .filterNot(_.key.head == NonWitnessUTXOKeyId.byte)
+              .filterNot(_.key.head == WitnessUTXOKeyId.byte) :+ WitnessUTXO(
+              out)
+          } else {
+            elements.filterNot(_.key.head == NonWitnessUTXOKeyId.byte)
+          }
+        } else {
+          elements
+        }
+      }
+      InputPSBTMap(newElements)
+    }
+  }
+}
+
+object InputPSBTMap extends PSBTMapFactory[InputPSBTRecord, InputPSBTMap] {
+  import org.bitcoins.core.psbt.InputPSBTRecord._
+
+  /** Constructs a finalized InputPSBTMap from a UTXOSpendingInfoFull,
+    * the corresponding PSBT's unsigned transaction, and if this is
+    * a non-witness spend, the transaction being spent
+    */
+  def finalizedFromUTXOSpendingInfo(
+      spendingInfo: UTXOSpendingInfoFull,
+      unsignedTx: Transaction,
+      nonWitnessTxOpt: Option[Transaction])(
+      implicit ec: ExecutionContext): Future[InputPSBTMap] = {
+    val sigComponentF = BitcoinSig
```

Comment:
There is way to much stuff in this file. This should be 3 or 4 files

Reply:
@nkohen